### PR TITLE
[WIP] #996 Add two-factor authentication

### PR DESCRIPTION
**Pull request by asfaltboy**
EuroPython 2018 collaborators working on #996

### CODE REVIEW

**Comment by di** [comment]
Thanks for the PR! I think we'll probably want to hold off on merging this until we have the rest of #996 implemented as well.

**Comment by Sparkycz** [comment]
Printscreen of the verification page http://prntscr.com/kc4x8e

**Comment by Sparkycz** [comment]
@asfaltboy Could you please edit the title of the PR? This one is a little bit confusing.. Maybe to `#996 add two-factor authentication`

**Comment by asfaltboy** [comment]
Good point @Sparkycz done that!

**Comment by Sparkycz** [comment]
I think we should add one column more into the table AccountUser to indicate that the user has enabled two-factor auth.

Something like..
```
op.add_column(
    "accounts_user", sa.Column("2fa_enable", sa.Boolean(), nullable=False, default=False)
)
```

However, i'm still working on the login presenters..

**Comment by sergeykolosov** [comment]
@Sparkycz I'd like to make a number of clean-ups, commit squashes, etc, which basically involves rewriting branch history. Please notify me when it's an appropriate moment to do so, so that it won't conflict with your work on presenters.

**Comment by sergeykolosov** [comment]
Updated the branch as follows:
* fixed requirements ordering;
* renamed OTP secret to TOTP secret (to avoid collisions as soon as other OTP methods are added);
* fixed migration not being regenerated after field length change;
* reordered and squashed commits into logically consistent groups;
* implemented TOTP primitives and documented design decisions around those (see docstings);
* covered TOTP primitives with basic unit-tests.

@nlhkabu suggested teaming up with others involved in #996 to avoid putting effort into duplicate artifacts.

**Comment by sergeykolosov** [comment]
@Sparkycz please address unittest & linting issues.

**Comment by pradyunsg** [comment]
Could someone add the EuroPython label on this? :)

**Comment by sergeykolosov** [comment]
Those who had already applied DB migrations, before pulling the updated branch, make sure to downgrade first:
```
docker-compose run web python -m warehouse db downgrade 263b37556f74-1
```

**Comment by dstufft** [comment]
FWIW, we do ``git merge --squash``, so you don't have to worry at all about keeping a clean history.

File: warehouse/utils/otp.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    https://tools.ietf.org/html/rfc4226#section-4
+
+    Given the base32 encoding, it is represented as a 32-character string, i.e.:
+    5 bits per character * 32 characters = 160 bits.
+
+    TODO:
+    consider using os.urandom(); this, however, is likely to cause
+    compatibility issues unless base32 vocabulary is used.
+    """
+    return pyotp.random_base32(length=32)
```
Comment (dstufft) [inline warehouse/utils/otp.py]: Looking at this function, it's not actually being base32 encoded, it's just generating a 16 length random string which just happens to also be valid base32. If this just needs a be a base32 encoded string, then I'd prefer to do something like:

```python
import base64
import os

def generate_totp_secret():
    return base64.b32encode(os.urandom(32)).decode("ascii")
```

This matches how we've done things in ``warehouse/utils/crypto.py``.

File: warehouse/utils/otp.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        * 30-second interval
+    """
+    return pyotp.TOTP(secret).provisioning_uri(username, issuer_name=issuer_name)
+
+
+def verify_totp(secret, value, valid_window=1):
+    """
+    Verifies a given TOTP-secret and value.
+
+    The *valid_window* argument value is intentionally chosen to be non-zero
```
Comment (dstufft) [inline warehouse/utils/otp.py]: Can we add a note here as to what exactly the valid window translate into? Is it an extra second? 6?, 30?

**Comment by Sparkycz** [comment]
I amended fix of the routes test into the commit with template. However there is still output from `black` linter (would reformat .. ) but i wouldn't like to reformat before the end of the changes in acounts/views.py

Meanwhile i'm finishing first version of the presenters (1-2 unit-tests left) despite We'll have to discuss the behavior of sessions etc..

**Comment by Sparkycz** [comment]
i guess you'll have a lot of requests for changes but somewhere we've gotta start ;-) 

pls check todos:
1. add enable/disable option of 2FA to profile settings
2. implement calling of utils/otp functions to user_service

**Comment by woodruffw** [comment]
I've restarted work on this in https://github.com/pypa/warehouse/compare/master...trailofbits:tob/996-add-2fa-support. Will open a PR shortly so that others can review progress in real time.

**Comment by di** [comment]
Closing in favor of #5567, but thank you for your contribution!

**Pull request closed without merging**